### PR TITLE
feat: rendering component instances based on array prop

### LINKED ISSUE #2708: Support for array prop + render each and extracting data from items
## What Feature Can Be Enhanced?
We already have an array type for props, and we also have a logic for rendering _n_ times based on an array prop. But extracting the data from each items seems not working correctly.

We want an element to be rendered _n_ times while also showing the data from each items.

### Bug
We can see here that the label from each array item is not being displayed
image
image

### PR DESCRIPTION
**Pull request by jethrocabaluna**
## Description

This is for an attempt to copy this card grid component in this sample [e-commerce website ](https://demo.vercel.store/search?sort=trending-desc)
image

### Implementation
Added `childMapperPropKey` and `childMapperComponent` to the Element model
- `childMapperPropKey` should be set to the state or component prop, which ideally should be an array when evaluated
- `childMapperComponent` will be rendered _n_ times based on the evaluated `childMapperPropKey`

The `childMapperComponent` instances are rendered as the last children for now. We will implement later on that will allow the user to move the whole rendered instances around under the parent element.

The array item data can be accessed by `{{props.[propName]}}` inside the component. Basically its like this - 
```
<Container>
  {products.map((product) => (
    <Card name={product.name} price={product.price}>
      <Image src={product.imageUrl} />
    </Card>)
  )}
</Container>
```

The tree node for each rendered instances will also be added in the tree view in the left panel

## Video or Image

Using the child mapper fields in the `AntDesignRow` element to render the _Column Product_ component instances based on the products data fetched after clicking the button

https://github.com/codelab-app/platform/assets/27695022/82ce3581-b69c-4fd7-b0f5-4b2286994f3e

<br />

Using the child mapper inside a _Products_ component which is also the same as above, but is just made into a pure custom component. And the products are fetched on the pre-render action on the _Products_ instance

https://github.com/codelab-app/platform/assets/27695022/d169352d-5331-4559-9b3c-09f517503111

## Related Issue(s)

Fixes #2708

### CODE REVIEW

**Comment by webberwang** [comment]
- I think each product instance should only contain 1 set of fields, the array should be an array of product instances, so the tree view on the left should have n instances
- would this affect implementation or just the UI display?
- the reason being is the the array data should be at the level right away the instance of products, then the instance product is rendered n times based on the array data
- if we use resource endpoint data binding, say some graphql endpoint that gets us array data, we would be able to do render for each on the array with the instance n times

**Comment by jethrocabaluna** [comment]
> * I think each product instance should only contain 1 set of fields, the array should be an array of product instances, so the tree view on the left should have n instances
> * would this affect implementation or just the UI display?
> * the reason being is the the array data should be at the level right away the instance of products, then the instance product is rendered n times based on the array data
> * if we use resource endpoint data binding, say some graphql endpoint that gets us array data, we would be able to do render for each on the array with the instance n times

hmm yeah i think in this case
- we will need a new type that just accepts data (any kind) from props, state, api response, or expression
- _products_ instance has a field that just accepts the data, ideally an array of data, could be from the response of external APIs or just passed down from the props or state
- _products_ component has a child _product_ instance where the `renderForEachPropKey` value is that new field
- then duplicate the _product_ instance _n_ times in the tree view in the left panel based on the length of the array
  - although in the main tree view it will just show the _products_ instance

I think this will also work on non-component approach where we just create the grid element in the main tree and card element that has `renderForEachPropKey` value with any array data (from external API response or props or state)

The looping implementation in this PR should work in those cases but will need to
- update so that it also duplicates the element _n_ times in the tree view
- support a new type to accept any kind of data
  - or maybe we can make it so we can set its expected type e.g. `{ name: string, price: number }` or just a `string`
  - note this is different from our any existing types like interfaceTypes or the primitive types, because those types will force you to set the values. We want something to just pass a data to a component

**Comment by jethrocabaluna** [comment]
I will have to update this PR again once this  https://github.com/codelab-app/platform/pull/2732 is merged since it will have many conflict and need some feature from it.

This works for accessing a response from an API endpoint after saving to `this.[propName]` and setting an expression to the ArrayType field of the _products_ component.

In this video, I have two components
- Products
  - no _Product_ instance
  - elements to render the product cards are made directly under this component
- New Products
  - has _Product_ instance
  - the elements to render the product card is inside the _Product_ component

The looping logic works on both approach. The array item data can be accessed via `loopProps` (can think of a better name)

https://github.com/codelab-app/platform/assets/27695022/cf35d32c-3085-4f73-9efa-2b1b6e8decaa

File: libs/frontend/domain/element/src/store/element.model.ts [hunk 1]
```diff
@@ -366,11 +367,27 @@ export class Element
...
+        this.parentComponent?.current.runtimeProp?.componentEvaluatedProps,
+      )
+
+      const value = evaluateRenderForEachPropKey(this, props)
+
+      if (Array.isArray(value) && value.length > 0) {
+        primaryTitle = `(x${value.length}) ${primaryTitle}`
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (webberwang) [reply libs/frontend/domain/element/src/store/element.model.ts:385]: yeah sure we will revisit

**Pull request merged**